File: 23-merge-k-sorted-lists/merge-k-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    const queue = new MinPriorityQueue({ priority: x => x.val })

    for (head of lists) {
        if (head) {
        queue.enqueue(head)
        }
    }
    const dummy = new ListNode(0)
    let tail = dummy

    while(!queue.isEmpty()) {
        const node = queue.dequeue().element

        tail.next = new ListNode(node.val)
        tail = tail.next

        if (node.next) {
            queue.enqueue(node.next)
        }
    }

    return dummy.next
};

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists or len(lists) == 0:
            return None

        while len(lists) > 1:
            merged = []

            for i in range(0, len(lists), 2):
                l1 = lists.pop()
                l2 = lists.pop() if len(lists) > 0 else None
                merged.append(self.mergeList(l1, l2))
            
            lists = merged

        return lists[0]
        
    

    def mergeList(self, l1, l2):
        dummy = ListNode(0)
        tail = dummy

        while l1 and l2:
            if l1.val < l2.val:
                tail.next = ListNode(l1.val)
                l1 = l1.next
            else:
                tail.next = ListNode(l2.val)
                l2 = l2.next
            tail = tail.next
        
        tail.next = l1 or l2

        return dummy.next
